### PR TITLE
Attach decorator-level options to parameterized functions

### PR DESCRIPTION
**Pull request by gongy**
Look up and execute a parametrized function that is published by another workspace.

Syntax:

```python
modal.Cls.lookup(
  workspace="modal-labs", "vllm-inference", "Model"
).with_options(
  volumes={"/data": Volume.lookup("my-llama-weights")},
  allow_concurrent_inputs=10,
  concurrency_limit=3,
) # Attachs a api_pb2.FunctionOptions object
```

### CODE REVIEW

File: modal/cls.py [hunk 1]
```diff
@@ -160,6 +169,11 @@ def _get_metadata(self) -> api_pb2.ClassHandleMetadata:
             )
         return class_handle_metadata
 
+    def with_options(self, **kwargs: Any) -> "_Cls":
+        self._options = _parse_function_options(**kwargs)
```
Comment (erikbern) [inline modal/cls.py:173]: might be better for documentation to not use **kwargs here

File: modal/cls.py [hunk 2]
```diff
@@ -160,6 +169,11 @@ def _get_metadata(self) -> api_pb2.ClassHandleMetadata:
             )
         return class_handle_metadata
 
+    def with_options(self, **kwargs: Any) -> "_Cls":
```
Comment (erikbern) [inline modal/cls.py:172]: might be better for documentation to not use **kwargs here

File: modal/cls.py (hunk 1 above)
Comment (erikbern) [inline modal/cls.py:173]: you don't want to mutate the object here right? you want to return a new object with this set

File: modal_proto/api.proto [hunk 3]
```diff
@@ -685,6 +685,19 @@ message CheckpointInfo {
   string checkpoint_id = 3;
 }
 
+message FunctionOptions {
```
Comment (erikbern) [inline modal_proto/api.proto:688]: can you use the same `message Function` for this?

File: modal/cls.py (hunk 1 above)
Comment (gongy) [reply modal/cls.py:173]: i'm worried if I have the wrong approach

```python

    
def _initialize_from_other(self, other: "_Cls"):
    self._user_cls = other._user_cls
    self._functions = other._functions
    self._options = other._options
    self._callables = other._callables
    self._output_mgr: Optional[OutputManager] = other._output_mgr
    
 # and something like this
 
 
@classmethod
def from_existing(existing_cls):
    """Produces a new Cls from an existing Cls"""

    cls = _Cls._from_loader(
        existing_cls,
        existing_cls._load,
        existing_cls._rep,
        existing_cls._is_another_app,
        existing_cls._preload,
        existing_cls._hydrate_lazily,
        existing_cls._deps,
    )
    cls._initialize_from_existing(existing_cls)
        
 
 

```

Comment (erikbern) [reply modal/cls.py:173]: not following – i'm saying `with_options` should not mutate `self` but return a new object instead

**Comment by luiscape** [comment]
What do you think of adopting the same syntax Docker Hub, Hugging Face Hub et al use with the namespace being determined by a identifier before a `/`?

```
```

**Comment by gongy** [comment]
Definitely want to go in that direction @luiscape. Aiming to add this interface in an open-ended way that lets us move to a cleaner format later.

File: modal/cls.py [hunk 4]
```diff
@@ -227,6 +252,52 @@ async def _load_remote(obj: _Object, resolver: Resolver, existing_object_id: Opt
...
+        retries: Optional[Union[int, Retries]] = None,
+        timeout: Optional[int] = None,
+        concurrency_limit: Optional[int] = None,
+        allow_concurrent_inputs: Optional[int] = None,
+        container_idle_timeout: Optional[int] = None,
+        keep_warm: Optional[int] = None,
+        allow_background_volume_commits: bool = False,
+    ) -> "_Cls":
+        # TODO(gongy): figure out a cleaner way to clone a _Cls
+        cls = _Cls._from_loader(self, self._load, self._rep)
```
Comment (gongy) [inline modal/cls.py:270]: This clone needs review

Comment (erikbern) [reply modal/cls.py:270]: yeah this seems a bit weird. I think you're assuming that `cls` is hydrated so you could just skip `_from_loader`.

You could probably add a simple clone method on the `_Object` base class that does `_Object.__new__(cls)` and then calls `_initialize_from_other`

File: modal/object.py [hunk 5]
```diff
@@ -117,6 +121,17 @@ def _init_from_other(self, other: O):
         # Transient use case, see Dict, Queue, and SharedVolume
         self._init(other._rep, other._load, other._is_another_app, other._preload)
 
+    @classmethod
+    def _from_other(cls: Type[O], other: "_Object"):
```
Comment (erikbern) [inline modal/object.py:125]: minor thing but you might want to add `-> O` to the signature

Comment (erikbern) [reply modal/object.py:125]: actually why is this a classmethod? it probably makes more sense as an instnace method:

```python
def clone(self) -> O:
   ...

**Comment by gongy** [comment]
@prbot approve

**Review by erikbern — APPROVED** [review]
🚢

**Comment by gongy** [comment]
It inherits the options set when the class was originally wrapped.

**Pull request merged**